### PR TITLE
ref(escalating-issues): Remove cache from escalating issue detection

### PR DESCRIPTION
**Pull request by jangjodi**
Remove cache check from escalating issue detection since nodestore already caches.

This is a followup to https://github.com/getsentry/sentry/pull/47843

### CODE REVIEW

File: src/sentry/tasks/post_process.py [hunk 1]
```diff
@@ -674,10 +674,7 @@ def process_snoozes(job: PostProcessJob) -> None:
         and group.status == GroupStatus.IGNORED
         and group.substatus == GroupSubStatus.UNTIL_ESCALATING
     ):
-        if not is_escalating(group):
-            job["has_reappeared"] = True
-            return
-        job["has_reappeared"] = False
+        job["has_reappeared"] = is_escalating(group)
```
Comment (jangjodi) [inline src/sentry/tasks/post_process.py:677]: fix: we should've been setting `job["has_reappeared"] = True` if the job is escalating since it transitions to `escalating`, not the other way around

File: tests/sentry/issues/test_escalating.py [hunk 2]
```diff
@@ -246,11 +235,16 @@ def test_daily_count_query(self) -> None:
...
             group = event.group
 
-        # The group had 2 events yesterday
-        one_day_ago_mins = 24 * 60
+        # The group had 2 events yesterday, 10 mins before midnight
+        # Tests that events are aggregated in the daily count query by date, not by 24 hr periods
+        now = datetime.now()
+        mins_since_ten_mins_before_today = (
+            int((now - now.replace(hour=0, minute=0, second=0, microsecond=0)).total_seconds() / 60)
+            + 10
```
Comment (jangjodi) [inline tests/sentry/issues/test_escalating.py:243]: modified this test so that yesterday's events occurred 10 mins before today, instead of 24 hours ago

File: src/sentry/issues/escalating_group_forecast.py [hunk 3]
```diff
@@ -40,11 +40,11 @@ def save(self) -> None:
         nodestore.set(
             self.build_storage_identifier(self.project_id, self.group_id),
             self.to_dict(),
-            ttl=timedelta(TWO_WEEKS_IN_DAYS_TTL),
+            ttl=timedelta(ONE_WEEKS_IN_DAYS_TTL),
```
Comment (wedamija) [inline src/sentry/issues/escalating_group_forecast.py:43]: It might be wise to keep this at 2 weeks. Then if the job is delayed at least we have data in place.

I'd also recommend renaming `TWO_WEEKS_IN_DAYS_TTL` to `GROUP_FORECAST_TTL`, that way you don't need to rename it whenever the value change.

File: src/sentry/issues/escalating.py [hunk 4]
```diff
@@ -214,26 +213,17 @@ def get_group_daily_count(project_id: int, group_id: int) -> int:
...
-        # Set the cache to be valid until the next weekly escalating forecast task is run
-        forecast_cache_duration = (
-            (escalating_forecast[0] + timedelta(days=ONE_WEEK_DURATION)).date() - date_now
-        ).total_seconds()
-        cache.set(forecast_cache_key, escalating_forecast, forecast_cache_duration)
+    escalating_forecast = EscalatingGroupForecast.fetch(group.project.id, group.id)
+    # escalating_forecast = (
+    #     escalating_forecast.date_added,
+    #     escalating_forecast.forecast,
+    # )
```
Comment (wedamija) [inline src/sentry/issues/escalating.py:221]: Nit: Delete rather than keep commented out code around

**Review by wedamija — APPROVED** [review]

File: src/sentry/issues/escalating_group_forecast.py (hunk 3 above)
Comment (jangjodi) [reply src/sentry/issues/escalating_group_forecast.py]: good point, thanks for bringing this up

File: tests/sentry/issues/test_escalating.py [hunk 5]
```diff
@@ -246,11 +235,16 @@ def test_daily_count_query(self) -> None:
             )
             group = event.group
```
Comment (armenzg) [inline tests/sentry/issues/test_escalating.py:236]:
```suggestion
```

File: tests/sentry/issues/test_escalating.py [hunk 6]
```diff
@@ -246,11 +235,16 @@ def test_daily_count_query(self) -> None:
...
 
-        # The group had 2 events yesterday
-        one_day_ago_mins = 24 * 60
+        # The group had 2 events yesterday, 10 mins before midnight
+        # Tests that events are aggregated in the daily count query by date, not by 24 hr periods
+        now = datetime.now()
+        mins_since_ten_mins_before_today = (
```
Comment (armenzg) [inline tests/sentry/issues/test_escalating.py:241]: Create a variable for last times midnight and use that as the base variable from which to build other variables. You will need it for my comment about this being an intermittent test if the CI was to execute around UTC midnight.

**Review by armenzg — CHANGES_REQUESTED** [review]

**Comment by armenzg** [comment]
There are some changes in this PR (e.g. current logic fixes) that could be pulled out into its own PR since we want to start testing out escalating issues internally today.

File: tests/sentry/issues/test_escalating.py (hunk 6 above)
Comment (jangjodi) [reply tests/sentry/issues/test_escalating.py]: thank you for this suggestion!

File: src/sentry/issues/escalating.py [hunk 7]
```diff
@@ -193,47 +194,39 @@ def _extract_project_and_group_ids(groups: Sequence[Group]) -> Dict[int, List[in
...
-        ],
-    )
-    request = Request(dataset=Dataset.Events.value, app_id=IS_ESCALATING_REFERRER, query=query)
-    return int(raw_snql_query(request, referrer=IS_ESCALATING_REFERRER)["data"][0]["count()"])
+    key = f"daily-group-count:{project_id}:{group_id}"
+    daily_count = cache.get(key)
+
+    if daily_count is None:
+        today = datetime.now().date()
+        start_date = datetime.combine(today, datetime.min.time())
```
Comment (armenzg) [inline src/sentry/issues/escalating.py:202]: I would personally name the variable `midnight` to help the read.

TIL:
```
python -c "from datetime import datetime; print(datetime.combine(datetime.now().date(), datetime.min.time()))"
2023-04-26 00:00:00
```

File: src/sentry/issues/escalating.py [hunk 8]
```diff
@@ -41,6 +41,7 @@
 BUCKETS_PER_GROUP = 7 * 24
 ONE_WEEK_DURATION = 7
 IS_ESCALATING_REFERRER = "sentry.issues.escalating.is_escalating"
+GROUP_DAILY_COUNT_DURATION = 60
```
Comment (armenzg) [inline src/sentry/issues/escalating.py:44]:
```suggestion
GROUP_DAILY_COUNT_TTL = 60
```

A personal preference: TTL helps me know that it is for the purpose of caching. Up to you if you want to take it.

File: src/sentry/issues/escalating_group_forecast.py [hunk 9]
```diff
@@ -55,6 +55,13 @@ def fetch(cls, project_id: int, group_id: int) -> Optional[EscalatingGroupForeca
             date_added=datetime.now(),
         )
 
+    @classmethod
+    def fetch_todays_forecast(cls, project_id: int, group_id: int) -> int:
+        date_now = datetime.now().date()
+        escalating_forecast = EscalatingGroupForecast.fetch(project_id, group_id)
+        forecast_today_index = (date_now - escalating_forecast.date_added.date()).days
```
Comment (armenzg) [inline src/sentry/issues/escalating_group_forecast.py:62]: Is this not causing a bug in the tests? Am I doing something wrong?

```
> from datetime import datetime
> date_now = datetime.now().date()  # Line 60
> print(date_now)
2023-04-26
> date_now - datetime.now()
TypeError: unsupported operand type(s) for -: 'datetime.date' and 'datetime.datetime'
> type(date_now)
datetime.date
> type(datetime.now())
datetime.datetime
```

Random typing thing I've noticed. Why is `datetime` different between the two classes?
https://github.com/getsentry/sentry/blob/82e7fe0bdfb482b11743c0ec0552d8ac02fbc2ad/src/sentry/issues/escalating_group_forecast.py#L24
https://github.com/getsentry/sentry/blob/82e7fe0bdfb482b11743c0ec0552d8ac02fbc2ad/src/sentry/issues/escalating_group_forecast.py#L37

Could you check if we have typing issues in here?

Here are some notes I was working on that may be useful in the future
----------------------------------------------------------------------

I can't, at first read, if there would be any sort of bug in here. You probably know most of what I will mention here, however, for people looking at this PR later it will be good background knowledge.

In the large majority of the cases, a group that is marked as "archived until escalating" will *NOT* have a group forecast because we will save storing 95% of all groups in the node store.

In such cases, we will fetch the default low ceiling 
https://github.com/getsentry/sentry/blob/82e7fe0bdfb482b11743c0ec0552d8ac02fbc2ad/src/sentry/issues/escalating_group_forecast.py#L51-L56

File: tests/sentry/issues/test_escalating.py [hunk 10]
```diff
@@ -241,22 +234,22 @@ def test_daily_count_query(self) -> None:
...
-                fingerprint="group-query", minutes_ago=one_day_ago_mins + i
-            )
-            group = event.group
+            # Event occured i hours after yesterday midnight
+            event = self._load_event_for_group(minutes_ago=mins_since_yesterday_midnight + i * 60)
 
         # The group has 1 event today
         for i in range(2, 1, -1):
-            event = self._load_event_for_group(fingerprint="group-query", minutes_ago=i)
+            event = self._load_event_for_group(minutes_ago=i)
```
Comment (armenzg) [inline tests/sentry/issues/test_escalating.py:252]: This would still suffer from intermittency issues but let's move it to another PR.
I think `_load_event_for_group` should have a parameter being the reference datetime to create an event from.

**Review by armenzg — COMMENTED** [review]
Only one of my comments I need an answer to in order to approve this.

I think there's a bug in `forecast_today_index = (date_now - escalating_forecast.date_added.date()).days`

File: src/sentry/issues/escalating_group_forecast.py (hunk 9 above)
Comment (jangjodi) [reply src/sentry/issues/escalating_group_forecast.py:62]: **Addressing `forecast_today_index = (date_now - escalating_forecast.date_added.date()).days`:**
- Both `date_now` and `escalating_forecast.date_added.date()` are of type `datetime.date()` so this issue does not occur (please let me know if this is not what you were referring to in your comment)

**Addressing the differing types for `date_added`**:
- Objects of type `datetime` are not serializable and thus cannot be stored in nodestore. I work around this by converting it to a float timestamp and then back to a datetime. This is how it is done for the `detection_time` field in  [IssueOccurrence](https://github.com/getsentry/sentry/blob/a506567bd35fd0cd4a7678080d3ff8418313b128/src/sentry/issues/issue_occurrence.py)

File: tests/sentry/issues/test_escalating.py (hunk 10 above)
Comment (jangjodi) [reply tests/sentry/issues/test_escalating.py:252]: Yes, I agree. I think I misunderstood when you brought this up earlier. I'll open a new PR for this

**Review by armenzg — APPROVED** [review]
